stdpopsim/models.py

```diff
@@ -137,6 +137,153 @@ def asdict(self):
             "migration_matrix": self.migration_matrix,
             "demographic_events": self.demographic_events}
 
+# ###############################################################
+# UNDER CONSTRUCTION - GROUND TRUTH N_t
+# ###############################################################
+
+    def population_size_trajectory(self, end, num_steps=10):
+        """
+        This function computes the defined effective population size
+        for each sub-population, for any demographic model.
+
+        `end` specifies at which generation to stop
+        collecting effective population sizes among subpopulations.
+        The sampling will always begin at generation 0
+
+        `num_steps` parameter will determine how many points from
+        0 to `end` to sample.
+
+        Returns a tuple where the first value
+        is the steps (in generations) and the second is a
+        num_steps-by-num_pops ndarray that will contain the effective
+        population sizes for each subpopulation individually, for each step,
+        as defined by the model which inherits from this class.
+        """
+
+        num_pops = self.num_pops()
+        N_t = np.zeros([num_steps, num_pops])
+        steps = np.linspace(0, end, num_steps)
+        for j, t in enumerate(steps):
+            N = self.pop_size_and_migration_at_t(t)["population_sizes"]
+            N_t[j] = N
+        return steps, N_t
+
+    def coalescence_rate_trajectory(self, end, num_samples, num_steps=10):
+        """
+        This function will calculate the ground truth
+        effective population size (Ne) for a population with
+        multiple sub-populations. This takes a
+        demographic model and returns a
+        function of time that gives haploid coalescence rate.
+
+        `num_samples` should be a list of the same length as the number
+        of populations, so that num_samples[j] is the number of
+        samples in subpopulation j.
+
+        `num_steps` parameter will determine how many points along the
+        time axis are returned.
+
+        Returns a tuple where the first value
+        is the steps (in generations) and the second is a numpy
+        ndarray that will contain the coalescence rate for
+        the entire population.
```

Comment:
"the second is an array whose `j`th element is the coalescence rate at time `steps[j]`, i.e., the probability per unit time that a randomly chosen pair of lineages from the original samples coalesces around that time, conditioned on the pair not having yet coalesced."